### DISCUSSION TITLE
Multilingual custom extension

### DISCUSSION THREAD

**Discussion by sfederici** [TurboWarp/extensions#1168] [Q&A]
Is it possible to add multilingual translations to custom turbowarp extensions? If so, how? Thanks in advance

**Comment by CubesterYT** [comment]
As in making extensions support other languages? Then yes, it's possible, as of recently, you can use `Scratch.translate("STRING")` on any strings to be able to get automatically translated by supporting transifex backed translations.

**Reply by VariedadesJuanYT** [reply]
y donde pongo ese codigo?

**Comment by GarboMuffin** [comment]
Documentation to be written on this soon

If you intend to submit the extensions here, it's easiest to just do this as our build server has various tricks:

```js
Scratch.translate("English version");

// Or with context:
Scratch.translate({
  default: "English version",
  description: "This text appears as 'developer comments' in Transifex"
});
```

If you want to do it manually for a custom extension:

```js
// Instead of hardcoding English/Italian/whatever strings, use this:
Scratch.translate({
  id: "stringid", // just needs to be unique within the extension
  default: "English version"
});

// To tell Scratch.translate about the different languages, do this once at the start, before calling Scratch.translate():
Scratch.translate.setup({
  it: {
    stringid: "Italian version"
  },
  es: {
    stringid: "Spanish version"
  }
});
```

**Reply by sfederici** [reply]
> If you want to do it manually for a custom extension:

I tried replacing

`text: "forever if [CONDITION]",`

with

`text: Scratch.translate("foreverif") + " [CONDITION]",`

then added 

```
// Instead of hardcoding English/Italian/whatever strings, use this:
Scratch.translate({
	id: "foreverif", // just needs to be unique within the extension
	default: "forever if"
});

// To tell Scratch.translate about the different languages, do this once at the start, before calling Scratch.translate():
Scratch.translate.setup({
	it: {
		"foreverif": "per sempre quando"
	}
});
```

outside the sandboxed extension, that is after "`})(Scratch)`;". The `foreverif` id doesn't get translated. If I put the code inside the class I get a syntax error. I'm totally new at using Scratch.translate. Any hint? 

BTW: [foreverif.js.zip](https://github.com/TurboWarp/extensions/files/13474535/foreverif.js.zip) is the "translated" extension

**Reply by CST1229** [reply]
I think you have to put the `Scratch.translate.setup()` call before the `Scratch.extensions.register()` (and inside the IIFE), and probably also include the `[CONDITION]` as part of the string. Also, that example Scratch.translate call there in the second code block should be in the first, like
```
text: Scratch.translate({
	id: "foreverif",
	default: "forever if [CONDITION]"
});
```

**Reply by sfederici** [reply]
Oh, I see. That makes sense, even if it doesn't explain to me how 

`Scratch.translate("English version"); `

would work then.

Anyway, I don't think `[CONDITION]` should be included. It would be really weird to me if it should. I'm going to test and I will let you know.

**Reply by sfederici** [reply]
Yes, it works. I have put

```
Scratch.translate.setup({
	it: {
		"foreverif": "per sempre quando"
	}
});
```

at the beginning of the IIFE and then I used

`text: Scratch.translate({ id: "foreverif", default: "forever if"}) + " [CONDITION]",
`

inside the new block definition.

Now, when I load the test pong.sb3 file, everything works, even changing the language from en to it.

**Reply by CST1229** [reply]
>Anyway, I don't think `[CONDITION]` should be included. It would be really weird to me if it should.

Some languages may reorder inputs, or put text at the end of the block, and blocks can also have multiple inputs. Both the vanilla Scratch blocks and the vanilla extension blocks have the inputs in the string for this reason.

**Reply by sfederici** [reply]
Yes, you are right. I will give it a try

**Comment by sfederici** [comment]
I would gladly submit the extension, but the guidelines say that submitted extensions should be of general interest. I don't know if adding back the "forever if" block of Scratch 1.4 could be considered of general interest?

**Reply by GarboMuffin** [reply]
Extension review is currently very sporadic and slow. You can try if you want but I would suggest just doing your manual approach and let submitting it here be a secondary side thing

**Reply by sfederici** [reply]
Ok thanks

**Comment by DogLover8425** [comment]
How come this doesn't work:
Scratch.translate.setup({"es": {"_moo": "MOOEA", "_I am a test!": "TESTA"}});
class Test {
getInfo() {
        return {
            id: "test",
            name: Scratch.translate("I am a test!"),
            blocks: [
                {
                    opcode: "Test block",
                    blockType: Scratch.BlockType.COMMAND,
                    text: Scratch.translate("moo")
                }
            ]
        };
    }
}

Scratch.extensions.register(new Test());

Or this: Scratch.translate.setup({"es": {"_moo": "MOOEA", "_I am a test!": "TESTA"}});
class Test {
    getInfo() {
        return {
            id: "test",
            name: Scratch.translate("I am a test!"),
            blocks: [
                {
                    opcode: "Test block",
                    blockType: Scratch.BlockType.COMMAND,
                    text: Scratch.translate("moo")
                }
            ]
        };
    }
}

Scratch.extensions.register(new Test());


Even if you switch to Spanish Latin America or Spanish Spain, it doesn't work.

**Reply by GarboMuffin** [reply]
i think what you're seeing is that we never bothered to make translate work in sandboxed extensions

**Reply by GarboMuffin** [reply]
honestly i don't know if sandboxed extensions will even exist in the future so it's not a high priority

**Reply by DogLover8425** [reply]
Why don't you just fix it? Besides the fact that have to, you know, GO THROUGH THE ENTIRE PROJECT AND FIND THE SINGLE LINE WHERE IT HAPPENS.
